### PR TITLE
Translation pipeline

### LINKED ISSUE #619: Implement automatic download and pull request creation for fully translated content from Crowdin
When translation of a markdown file on Crowdin is completed, this should automatically be retrieved and included in a pull request to the `develop` branch for review.
 
Although Crowdin ships with GitHub integration, this is not suitable for our use case. We want to selectively include translations once they are available.  The included GitHub integration seems more designed for cases where all translation occurs before a deployment, and so it includes ALL translated files (including non/partially translated files which still contain source strings) in pull requests it makes. 

Currently, the only mechanism to be notified of a completed translation is via a webhook. This would require a server running on some platform to receive the notification (via an HTTP request), download the translated file, and open a pull request.

### LINKED ISSUE #620: Implement automatic upload of new content file on Develop to Crowdin
When new/updated content markdown files are merged onto `master`, these need to be uploaded to Crowdin. 

While Crowdin ships with GitHub integration, this is not suitable for this use case (see #619). Instead, this will probably need to be implemented as a Travis job that runs for new commits to `master`. Upload can be achieved through the Crowdin CLI (https://support.crowdin.com/cli-tool/).

### LINKED ISSUE #621: Implement automatic update of .po file
The .po file needs to be updated with all static strings for translation, by running the `python manage.py makemessages`. This needs to be run upon any change to templates or database content.

This process should be automated on Travis to run when any such files are updated on `develop`.

### LINKED ISSUE #622: Implement automatic upload to Crowdin of .po file(s) on develop branch
Any changes to any .po files (probably updated by Travis when any static translatable strings change, see #621) should be synchronised to the Crowdin server.

Any new translations completed on Crowdin to those files should trigger a pull request back into `develop` with the translated versions.

### LINKED ISSUE #705: Implement in-context localisation for CS Unplugged
See https://support.crowdin.com/in-context-localization/

### PR DESCRIPTION
**Pull request by jordangriffiths01**
Implement full translation pipeline for CS Unplugged, using an automation bot running on GCE.

Fixes #619 
Fixes #620 
Fixes #622 
Fixes #705

### CODE REVIEW

File: crowdin_content.yaml [hunk 1]
```diff
@@ -1,19 +1,42 @@
...
 
 files: [
     {
         source: "/csunplugged/topics/content/en/**/*.yaml",
-        translation: "/csunplugged/topics/content/%two_letters_code%/**/%original_file_name%",
-
+        translation: "/csunplugged/topics/content/%osx_locale%/**/%original_file_name%",
+        languages_mapping: {
+            osx_locale: {
+                "en-UD": "xx_LR",  # Pseudo language
```
Comment (JackMorganNZ) [inline crowdin_content.yaml]: Why is "xx_RL" not listed here?

File: Dockerfile-local [hunk 2]
```diff
@@ -9,6 +9,7 @@ LABEL maintainer="csse-education-research@canterbury.ac.nz"
 ARG DEBIAN_FRONTEND=noninteractive
 
 ENV DJANGO_PRODUCTION=False
+ENV INCLUDE_INCONTEXT_L10N=True
```
Comment (JackMorganNZ) [inline Dockerfile-local]: What happens when you view in-context mode on a branch with changes yet to make it to Crowdin?

File: csu [hunk 3]
```diff
@@ -127,7 +127,7 @@ defhelp -dev flush 'Run Django flush command.'
 # Run Django makemigrations command
 dev_makemigrations() {
   echo "Creating database migrations..."
-  docker-compose exec django /docker_venv/bin/python3 ./manage.py makemigrations
+  docker-compose exec django /docker_venv/bin/python3 ./manage.py makemigrations --no-input
```
Comment (JackMorganNZ) [inline csu]: Why is this change needed?

File: csunplugged/config/settings/base.py [hunk 4]
```diff
@@ -102,10 +106,43 @@
...
     ("en", "English"),
 )
 
+if env.bool("INCLUDE_INCONTEXT_L10N", False):
+    EXTRA_LANGUAGES = [
+        (INCONTEXT_L10N_PSEUDOLANGUAGE, "In-context translations"),
+        (INCONTEXT_L10N_PSEUDOLANGUAGE_BIDI, "In-context translations (Bidi)"),
+    ]
+
+    LANGUAGES += tuple(EXTRA_LANGUAGES)
```
Comment (JackMorganNZ) [inline csunplugged/config/settings/base.py]: This line appears unused?

File: csunplugged/templates/base.html [hunk 5]
```diff
@@ -6,13 +6,20 @@
...
 
+
 <!DOCTYPE html>
 <html lang="{{ LANGUAGE_CODE }}" {% if current_language.bidi %}dir="rtl"{% endif %}>
   <head>
     <!-- Required meta tags -->
     <meta charset="utf-8">
     <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
-
+    {% if LANGUAGE_CODE == "xx-lr" or LANGUAGE_CODE == "yy-rl" %}
```
Comment (JackMorganNZ) [inline csunplugged/templates/base.html]: Do we want to display these in the language picker to make it easier for translators? Would be easy to do.

File: infrastructure/crowdin/crowdin_bot_python_package/crowdin_bot/api.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+import json
+import os
+import requests
+
+API_KEY = os.environ["CROWDIN_API_KEY"]
+API_URL = "https://api.crowdin.com/api/project/cs-unplugged/{method}"
+
+
+def api_call_text(method, **params):
+    """Call a given api method and return text content.
```
Comment (JackMorganNZ) [inline infrastructure/crowdin/crowdin_bot_python_package/crowdin_bot/api.py]: Missing argument `method`.

File: infrastructure/crowdin/crowdin_bot_python_package/crowdin_bot/api.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    params["key"] = API_KEY
+    response = requests.get(
+        API_URL.format(method=method),
+        params=params
+    )
+    return response.text
+
+
+def api_call_xml(method, **params):
+    """Call a given api method and return XML tree."""
```
Comment (JackMorganNZ) [inline infrastructure/crowdin/crowdin_bot_python_package/crowdin_bot/api.py]: Missing argument and returns in docstring.

File: infrastructure/crowdin/crowdin_bot_python_package/crowdin_bot/api.py [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    response_text = api_call_text(method, **params)
+    xml = lxml.etree.fromstring(response_text.encode())
+    return xml
+
+def api_call_json(method, **params):
+    """Call a given api method and return JSON dictionary."""
```
Comment (JackMorganNZ) [inline infrastructure/crowdin/crowdin_bot_python_package/crowdin_bot/api.py]: Missing argument and returns in docstring.

File: infrastructure/crowdin/crowdin_bot_python_package/crowdin_bot/get_complete_translations.py [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+import os
+import argparse
+
+from crowdin_bot import api
+
+SOURCE_LANGUAGE = "en"
+
+def get_language_info(language):
+    """Get xml tree from language info api call."""
```
Comment (JackMorganNZ) [inline infrastructure/crowdin/crowdin_bot_python_package/crowdin_bot/get_complete_translations.py]: Missing argument and returns in docstring.

File: infrastructure/crowdin/crowdin_bot_python_package/crowdin_bot/get_complete_translations.py [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+    """Return list of completely translated file paths in a given directory tree node.
+
+    Args:
+        item: (etree.Element): itemm node in language-status xml tree
+            (see https://support.crowdin.com/api/language-status/)
+        parent_path: (str) path to the translated file node (None if the current item is
+            the root of the directory tree).
+        csu_language_code: (str) Language code (in locale format) on CSU end
+            (may differ from crowdin language code according to language mapping
+            in yaml file)
```
Comment (JackMorganNZ) [inline infrastructure/crowdin/crowdin_bot_python_package/crowdin_bot/get_complete_translations.py]: Missing returns in docstring.

File: infrastructure/crowdin/crowdin_bot_python_package/crowdin_bot/get_crowdin_files.py [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+    return api.api_call_xml("info")
+
+
+def process_item(item, parent_path="/"):
+    """Return list of paths to all files under a given node
+
+    Args:
+        item: (etree.Element): itemm node in info xml tree
+            (see https://support.crowdin.com/api/info/)
+        parent_path: (str) path to the file/folder node "item"."""
```
Comment (JackMorganNZ) [inline infrastructure/crowdin/crowdin_bot_python_package/crowdin_bot/get_crowdin_files.py]: Missing returns in docstring.

File: infrastructure/crowdin/crowdin_bot_python_package/crowdin_bot/get_crowdin_files.py [hunk 12]
```diff
@@ -0,0 +1,39 @@
+"""Script to print list of all source file paths on crowdin."""
+
+import os
+from crowdin_bot import api
+
+
+def get_project_info():
+    """Get xml containing all crowdin files (from api)."""
```
Comment (JackMorganNZ) [inline infrastructure/crowdin/crowdin_bot_python_package/crowdin_bot/get_crowdin_files.py]: Missing return in docstring.

File: infrastructure/crowdin/crowdin_bot_python_package/crowdin_bot/get_crowdin_languages.py [hunk 13]
```diff
@@ -0,0 +1,21 @@
+"""Script to print list of all crowdin language codes for project."""
+
+from crowdin_bot import api
+
+NS_DICT = {
+    'ns': "urn:oasis:names:tc:xliff:document:1.2"
+}
+
+def get_project_languages():
+    """Get list of crowdin language codes."""
```
Comment (JackMorganNZ) [inline infrastructure/crowdin/crowdin_bot_python_package/crowdin_bot/get_crowdin_languages.py]: Missing return in docstring.

File: infrastructure/crowdin/crowdin_bot_python_package/crowdin_bot/modify_pseudo_translations.py [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+        xliff = etree.parse(f, XLIFF_PARSER)
+        xliff_trans = xliff.xpath(
+            './/ns:file/ns:body',
+            namespaces=NS_DICT
+        )[0]
+        return xliff_trans
+
+
+def get_verto_block_patterns():
+    """Get list of compiled regex patterns for verto block tags."""
```
Comment (JackMorganNZ) [inline infrastructure/crowdin/crowdin_bot_python_package/crowdin_bot/modify_pseudo_translations.py]: Missing return in docstring.

File: infrastructure/crowdin/crowdin_bot_scripts/crowdin-bot-config.sh [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+IN_CONTEXT_L10N_TARGET_BRANCH="develop"
+
+# Name of crowdin config file
+CROWDIN_CONFIG_FILE="crowdin_content.yaml"
+
+# Code for in-context localisation pseudo language on Crowdin
+CROWDIN_PSEUDO_LANGUAGE="en-UD"
+
+# Code for in-context localisation pseudo language on CS Unplugged
+CSU_PSEUDO_LANGUAGE="xx_LR"
```
Comment (JackMorganNZ) [inline infrastructure/crowdin/crowdin_bot_scripts/crowdin-bot-config.sh]: Does `XX_RL` not need to be listed here?

File: infrastructure/crowdin/crowdin_bot_scripts/crowdin-bot-push-source.sh [hunk 16]
```diff
@@ -0,0 +1,24 @@
+#!/usr/bin/env bash
+
+# Script to push source files from crowdin.
```
Comment (JackMorganNZ) [inline infrastructure/crowdin/crowdin_bot_scripts/crowdin-bot-push-source.sh]: Should this say "push source files **to** Crowdin"?

File: crowdin_content.yaml (hunk 1 above)
Comment (jordangriffiths01) [reply crowdin_content.yaml]: xx_RL is on the CSU end only, it's not a language crowin knows anything about (and is symlinked to xx_LR)

File: csunplugged/config/settings/base.py (hunk 4 above)
Comment (jordangriffiths01) [reply csunplugged/config/settings/base.py]: It's definitely used - the LANGUAGES variable defines the languages that are available to the end user. What were your thoughts?

File: csunplugged/templates/base.html (hunk 5 above)
Comment (jordangriffiths01) [reply csunplugged/templates/base.html]: They would be automatically, if the language picker was implemented. Perhaps instead of removing it completely, we should add it to develop, but wrapped in an `if` block that checks if the length of `LANGUAGES` is greater than 1?

File: Dockerfile-local (hunk 2 above)
Comment (jordangriffiths01) [reply Dockerfile-local]: As discussed offline, this isn't an issue. Content changes in a branch won't affect the in-context l10n features on that branch at all. Backend changes, while they might make the in-context page on that branch unusable until merged, won't break anything.

File: csu (hunk 3 above)
Comment (jordangriffiths01) [reply csu]: If you delete a field and add a similar field, django asks if you are renaming the column rather than deleting and adding. When this occurs due to deleting/adding a language, that's a manual Y/N answer for every field for every model. This prevents that user prompt.

File: infrastructure/crowdin/crowdin_bot_scripts/crowdin-bot-config.sh (hunk 15 above)
Comment (jordangriffiths01) [reply infrastructure/crowdin/crowdin_bot_scripts/crowdin-bot-config.sh]: Negative - YY_RL is permanently symlinked to XX_LR, and these scripts do not need to be aware of it's existence

**Comment by jordangriffiths01** [comment]
@JackMorganNZ Reference to fixing #621 was a typo, and has been fixed above. As discussed offline, docs for all i18n changes will come in a follow-up PR.
